ASoC: SOF: ipc4-topology: Adjust the params based on DAI formats

ujfalusi opened the pull request:
Currently we only check the bit depth value among to DAI formats, but other
parameters might be constant, like number of channels and/or rate.

In capture we use the fe params as a reference to find the format and blob
which should be used, but in the path we can have components which can
handle expanding/narrowing number of channels or do a resample.

In these cases the topology is expected to have 'fixed' parameter for
channels/rates/bit depth and the conversion to the fe format is going to
be done within the path.

In practice this patch fixes issues like:
All DMIC formats are fixed four channels
We have a component which converts the four channel to stereo
FE is opened with 2 channel

Even if we have the correct bit depth format and blob (for four channel) we
will still be looking for stereo configurations, which will fail.

Note: the adjustment of params have switched order with the checking of
single bit depth (needed for the NHLT blob fallback support). This change
is non function, just that if the sof_ipc4_narrow_params_to_format() would
fail, there is no point of checking the single bit depth.

ujfalusi committed:
ASoC: SOF: ipc4-topology: Adjust the params based on DAI formats

Currently we only check the bit depth value among to DAI formats, but other
parameters might be constant, like number of channels and/or rate.

In capture we use the fe params as a reference to find the format and blob
which should be used, but in the path we can have components which can
handle expanding/narrowing number of channels or do a resample.

In these cases the topology is expected to have 'fixed' parameter for
channels/rates/bit depth and the conversion to the fe format is going to
be done within the path.

In practice this patch fixes issues like:
All DMIC formats are fixed four channels
We have a component which converts the four channel to stereo
FE is opened with 2 channel

Even if we have the correct bit depth format and blob (for four channel) we
will still be looking for stereo configurations, which will fail.

Note: the adjustment of params have switched order with the checking of
single bit depth (needed for the NHLT blob fallback support). This change
is non function, just that if the sof_ipc4_narrow_params_to_format() would
fail, there is no point of checking the single bit depth.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-topology/pcm: Rename sof_ipc4_copier_is_single_format()

Rename the sof_ipc4_copier_is_single_format() to
sof_ipc4_copier_is_single_bitdepth() to clear the confusion of the use of
'format' when we are querying information on the bit depth.
Format is used to describe a combination of paramters (rate, channels,
sample format / bit depth).
Rename the flags used to store the result at the same time.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1582,55 +1582,92 @@ bool sof_ipc4_copier_is_single_bitdepth(struct snd_sof_dev *sdev,
 }
 
+static int
+sof_ipc4_narrow_params_to_format(struct snd_sof_dev *sdev,
```
"narrow params" does not read well, it can be a verb or an adjective. Did you mean 'adjust' or 'limit'?

ujfalusi replied:
Right, "adjust" is better, thanks

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1582,55 +1582,92 @@ bool sof_ipc4_copier_is_single_bitdepth(struct snd_sof_dev *sdev,
+	channels = SOF_IPC4_AUDIO_FORMAT_CFG_CHANNELS_COUNT(fmt->fmt_cfg);
+	valid_bits = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(fmt->fmt_cfg);
+
+	/* check if params in formts in topology are the same */
```
formats

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1582,55 +1582,92 @@ bool sof_ipc4_copier_is_single_bitdepth(struct snd_sof_dev *sdev,
+		return ret;
+
+	single_bitdepth = sof_ipc4_copier_is_single_bitdepth(sdev, pin_fmts,
+							     num_pin_fmts);
```
this looks weird, before this change the hw_params would be adjusted only if there's a single bit length, and now they are adjusted first and then you check for the single bitdepth.

Difficult to review or understand...

ujfalusi replied:
no, before this patch we only adjusted the FORMAT if the bit depth was constant, so we had single bit depth for all formats.
The change is that now we also adjust the RATE and CHANNELS at the same time as we did for the bit depth.

We adjust the dai_params based on the pin_fmts.

We still need the flag for the `snd_sof_get_nhlt_endpoint_data()` about the single bitdepth to be used to handle DMIC blob fallbacks.

In other words: there is no change in functionality with existing setups but we prepare for the future where the channels/rate is constant.

plbossart replied:
it'd probably be easier to review if you had an intermediate patch where you refactor the branch first with

````c
if (dir == SNDRV_PCM_STREAM_PLAYBACK) {
		pin_fmts = available_fmt->output_pin_fmts;
		num_pin_fmts = available_fmt->num_output_formats;
	} else {
		pin_fmts = available_fmt->input_pin_fmts;
		num_pin_fmts = available_fmt->num_input_formats;
	}
````

ujfalusi replied:
OK, let me do that

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1582,55 +1582,92 @@ bool sof_ipc4_copier_is_single_bitdepth(struct snd_sof_dev *sdev,
 	return true;
 }
 
+static int
```
@ujfalusi so we're relying on the topology to be always correct here isnt it? What will happen if the topology has only 4ch in all formats but the device only has 2 dmics?

plbossart replied:
Not following the concern @ranj063 ? A bad topology is generally the end of the story, even outside of format changes... We have no error concealment for problems in the topology, and the best solution is to fail big and fail early when a bad topology is encountered.
In this case trying to do a beamforming with only two mics is probably a mistake, isn't it?

ranj063 replied:
my concern is related to the number of physical mics on the device. If the topology only has 4ch format but the device only has 2 mics, should we use this topology? If not, how do we know to switch the topology with the 2ch format based on the number of mics on a chromebook. With the HDA topologies, this is easy because we have separate topologies for 2ch and 4ch dmics. But the I2S topologies dont have the distinction right?

ujfalusi replied:
@ranj063, if we have two dmics and the topology only have 4 channel only then it will not work no matter what you do. You would need 2 channel topology for the two DMICs, no?

I'm not sure if I follow your line of thought.

If we would 'trust' that the blob selection is the correct: the device have 2 DMICs, the loaded topology only have 4 channel configs. We cannot use the loaded topology with the device, right?

Isn't this the same constraint, but in reverse view? If there is no matching blob with the topology then it will not work.

We cannot 'switch' to a new topology, it is expected that the correct topology is loaded for the device, if not, things will not work and they did not worked in the past either?

ujfalusi replied:
@ranj063, in other words: if there is a single option for a parameter in topology for the DAI then we don't have much choice than using that single option. We cannot invent a new format in kernel, afaik.

plbossart replied:
I don't understand the concern @ranj063

Are you referring to the blobs defined with PDM0/PDM1, or the formats specified in the pipeline/components.
Those are two different parts IMHO.

singalsu replied:
We have the correct topology for microphones count, there can't be a conflict, kernel loads -2ch.tplg or -4ch.tplg based on NHLT information about microphones count. The topologies are generated for 2ch/4ch with help of NUM_DMICS macro, Then in dmic-generic.conf topology I use the same 

```
IncludeByKey.NUM_DMICS {
"2" {}
"4" {}
}
```
to select the dai-copier input format. Each channels count group have the format s16/24/32 variants. This is now working well in my tests.

singalsu replied:
To add, the above topology change is only needed when dai-copier's and host-copier's channels counts in connected pipelines can be different. Normally the channels counts are the same.

ujfalusi committed:
ASoC: SOF: ipc4-topology: Improve readability of sof_ipc4_prepare_dai_copier()

Remove the duplicated code paths to check for single bit depth and to
update the params with storing the parameters needed by the function and
have a single code section.

No functional change but the code is easier to follow.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v1:
- Use sof_ipc4_adjust_params_to_dai_format() as function name
- separate the readability change in sof_ipc4_prepare_dai_copier()
- Add note on the swap of params adjustment and querying of single bit depth (non functional change)

plbossart reviewed (approved):
LGTM, but @ranj063 and @bardliao  must review/approve.

singalsu reviewed (approved):
Thanks Peter this looks good and works perfectly in my test with 2ch and 4ch DMIC devices.

plbossart commented:
@ranj063 please review and merge if you're ok with this PR.
